webapp/tests/karma/ts/services/check-date.service.spec.ts

```diff
@@ -98,7 +102,122 @@ describe('CheckDateService', () => {
+    // client is one hour ahead of server
+    expect(telemetryService.record.args[0]).to.deep.equal(['client-date-offset', 60 * 60 * 1000]);
+  });
+
+  it('should not show the modal even when clock is wrong when param is not passed', async () => {
+    clock = sinon.useFakeTimers();
+    const responseDate = new Date();
+    responseDate.setHours(responseDate.getHours() - 5);
+    const check = service.check();
+    const res = httpMock.expectOne(matchInfoEndpoint);
+    res.flush('', { headers: { 'Date': responseDate.toISOString() } });
+
+    await check;
+    expect(modal.show.callCount).to.equal(0);
+    expect(telemetryService.record.callCount).to.equal(1);
+    // client is five hours ahead of server
+    expect(telemetryService.record.args[0]).to.deep.equal(['client-date-offset', 5 * 60 * 60 * 1000]);
+  });
+
+  describe('should keep re-checking on every try until one response from the server is received', () => {
+    it('when clock is off', async () => {
+      clock = sinon.useFakeTimers(1606230000000);
+
+      const check1 = service.check(true);
+      const res1 = httpMock.expectOne(matchInfoEndpoint);
+      res1.error(new ErrorEvent('Failed to fetch'));
+
+      await check1;
+      expect(modal.show.callCount).to.equal(0);
+      expect(telemetryService.record.callCount).to.equal(0);
+
+      const check2 = service.check(true);
+      const res2 = httpMock.expectOne(matchInfoEndpoint);
+      res2.error(new ErrorEvent('Failed to fetch'));
+
+      await check2;
+      expect(modal.show.callCount).to.equal(0);
+      expect(telemetryService.record.callCount).to.equal(0);
+
+      const check3 = service.check(true);
+      const res3 = httpMock.expectOne(matchInfoEndpoint);
+      res3.error(new ErrorEvent('Failed to fetch'));
+
+      await check3;
+      expect(modal.show.callCount).to.equal(0);
+      expect(telemetryService.record.callCount).to.equal(0);
+
+      const responseDate = new Date();
+      responseDate.setHours(responseDate.getHours() + 3);
+      const check4 = service.check(true);
+      const res4 = httpMock.expectOne(matchInfoEndpoint);
+      res4.flush('', { headers: { 'Date': responseDate.toISOString() } });
+
+      await check4;
+      expect(modal.show.callCount).to.equal(1);
+      expect(telemetryService.record.callCount).to.equal(1);
+      // client is 3 hours behind of server
+      expect(telemetryService.record.args[0]).to.deep.equal(['client-date-offset', - 3 * 60 * 60 * 1000]);
+
+      const check5 = service.check(true);
+      httpMock.expectNone(matchInfoEndpoint);
+
+      await check5;
+      expect(modal.show.callCount).to.equal(1);
+      expect(telemetryService.record.callCount).to.equal(1);
+
+      const check6 = service.check(true);
+      httpMock.expectNone(matchInfoEndpoint);
+
+      await check6;
+      expect(modal.show.callCount).to.equal(1);
+      expect(telemetryService.record.callCount).to.equal(1);
+    });
+
+    it('when clock is on', async () => {
```

Comment:
Use the same `should...` format here. I had to go back to the describe statement to understand what is being tested.